Piglet crashes after networked piggui disconnects

start piglet
start piggui with the piglet nodeid...
close piggui

piglet crashes with:
```
[2024-07-21T20:09:54Z INFO  piglet] Pin #26 Input level change: LevelChange { new_level: true, timestamp: 2024-07-21T20:09:54.478990Z }
[2024-07-21T20:09:55Z INFO  piglet] Pin #26 Input level change: LevelChange { new_level: false, timestamp: 2024-07-21T20:09:55.148646Z }
Error: timed out
make: *** [run-piglet] Error 1
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -44,6 +44,8 @@ rppal = { version = "0.18.0", optional = true }
 # used in piglet only
 log = "0.4.22"
 env_logger = "0.11.4"
+service-manager = "0.7.1"
+sysinfo = "0.30.13"
 
 # used by piggui in GUI only
 iced = { version = "0.12.1", default-features = false, features = ["tokio", "debug", "canvas", "advanced"], optional = true }

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -36,7 +36,7 @@ rand = { version = "0.8.5", optional = true }
 iroh-net = { version = "0.19.0" }
 anyhow = { version = "1" }
 futures-lite = { version = "2.3" }
-tokio = { version = "1", default-features = false, features = ["time", "rt"] }
+tokio = { version = "1.39.1", default-features = false, features = ["time", "rt"] }
 
 # for interacting with GPIO on the Raspberry Pi
 rppal = { version = "0.18.0", optional = true }

--- a/src/piglet.rs
+++ b/src/piglet.rs
@@ -116,7 +116,7 @@ fn check_unique(exec_path: &Path) -> anyhow::Result<PathBuf> {
         if let Some(path) = process.exe() {
             let info_path = path.with_file_name("piglet.info");
             if info_path.exists() {
-                println!("You can use the following info to connect to this running instance:");
+                println!("You can use the following info to connect to it:");
                 let piglet_info = fs::read_to_string(info_path)?;
                 println!("{}", piglet_info);
             }

--- a/src/piglet.rs
+++ b/src/piglet.rs
@@ -98,6 +98,7 @@ fn check_unique(exec_path: &Path) -> anyhow::Result<PathBuf> {
         .context("Could not get exec file name")?
         .to_str()
         .context("Could not get exec file name")?;
+    let info_path = exec_path.with_file_name("piglet.info");
 
     let my_pid = process::id();
     let sys = System::new_all();
@@ -107,16 +108,19 @@ fn check_unique(exec_path: &Path) -> anyhow::Result<PathBuf> {
         .collect();
     if let Some(process) = instances.first() {
         println!(
-            "An instance of {exec_name} is already running with PID='{}' with Path='{}', started by user with {:?}",
+            "An instance of {exec_name} is already running with PID='{}',  Path='{}', started by user with {:?}",
             process.pid(),
             process.exe().context("Could not get path to the running process instance")?.display(),
             process.user_id().context("Could not get the User ID of user who started the running process instance")?,
         );
 
+        println!("You can use the following info to connect to this running instance:");
+        let piglet_info = fs::read_to_string(info_path)?;
+        println!("{}", piglet_info);
+
         exit(1);
     }
 
-    let info_path = exec_path.with_file_name("piglet.info");
     // remove any leftover file from a previous execution - ignore any failure
     let _ = fs::remove_file(&info_path);
 
@@ -236,7 +240,7 @@ async fn listen(info_path: &Path, mut hardware: impl Hardware) -> anyhow::Result
     if let Some(connecting) = endpoint.accept().await {
         let connection = connecting.await?;
         let node_id = iroh_net::endpoint::get_remote_node_id(&connection)?;
-        info!("new connection from {node_id}",);
+        info!("New connection from nodeid: '{node_id}'",);
 
         let mut gui_sender = connection.open_uni().await?;
 

--- a/src/piglet.rs
+++ b/src/piglet.rs
@@ -34,15 +34,15 @@ const SERVICE_NAME: &str = "net.mackenzie-serres.pigg.piglet";
 /// over the network.
 #[tokio::main]
 async fn main() -> anyhow::Result<()> {
-    let hw = init()?;
-    listen(hw).await
-}
-
-/// Do initialization of logging, get access to local hardware and apply any config from a local
-/// file that maybe specified on the command line
-fn init() -> anyhow::Result<impl Hardware> {
     let matches = get_matches();
 
+    manage_service(&matches)?;
+
+    run_service(&matches).await
+}
+
+/// Handle any service installation or uninstallation tasks
+fn manage_service(matches: &ArgMatches) -> anyhow::Result<()> {
     let service_name: ServiceLabel = SERVICE_NAME.parse().unwrap();
 
     if matches.get_flag("uninstall") {
@@ -65,7 +65,13 @@ fn init() -> anyhow::Result<impl Hardware> {
         exit(0);
     };
 
-    setup_logging(&matches);
+    Ok(())
+}
+
+/// Run piglet as a service - this could be interactively by a user in foreground or started
+/// by the system as a user service, in background - use logging for output from here on
+async fn run_service(matches: &ArgMatches) -> anyhow::Result<()> {
+    setup_logging(matches);
 
     let mut hw = hw::get();
     info!("\n{}", hw.description().unwrap().details);
@@ -81,7 +87,8 @@ fn init() -> anyhow::Result<impl Hardware> {
         trace!("Configuration applied to hardware");
     };
 
-    Ok(hw)
+    // Then listen for remote connections and "serve" them
+    listen(hw).await
 }
 
 /// CHeck that this is the only instance of piglet running, both user process or system process
@@ -170,22 +177,30 @@ fn get_matches() -> ArgMatches {
     app.get_matches()
 }
 
-/// Listen for an incoming iroh-net connection
+/// Listen for an incoming iroh-net connection and apply any config changes received, and
+/// send to GUI over the connection any input level changes.
+/// This is adapted from the iroh-net example with help from the iroh community
 async fn listen(mut hardware: impl Hardware) -> anyhow::Result<()> {
     let secret_key = SecretKey::generate();
 
-    // Build a `Endpoint`, which uses PublicKeys as node identifiers, uses QUIC for directly connecting to other nodes, and uses the relay protocol and relay servers to holepunch direct connections between nodes when there are NATs or firewalls preventing direct connections. If no direct connection can be made, packets are relayed over the relay servers.
+    // Build a `Endpoint`, which uses PublicKeys as node identifiers, uses QUIC for directly
+    // connecting to other nodes, and uses the relay protocol and relay servers to holepunch direct
+    // connections between nodes when there are NATs or firewalls preventing direct connections.
+    // If no direct connection can be made, packets are relayed over the relay servers.
     let endpoint = Endpoint::builder()
-        // The secret key is used to authenticate with other nodes. The PublicKey portion of this secret key is how we identify nodes, often referred to as the `node_id` in our codebase.
+        // The secret key is used to authenticate with other nodes.
+        // The PublicKey portion of this secret key is how we identify nodes, often referred
+        // to as the `node_id` in our codebase.
         .secret_key(secret_key)
         // set the ALPN protocols this endpoint will accept on incoming connections
         .alpns(vec![PIGLET_ALPN.to_vec()])
         // `RelayMode::Default` means that we will use the default relay servers to holepunch and relay.
         // Use `RelayMode::Custom` to pass in a `RelayMap` with custom relay urls.
         // Use `RelayMode::Disable` to disable holepunching and relaying over HTTPS
-        // If you want to experiment with relaying using your own relay server, you must pass in the same custom relay url to both the `listen` code AND the `connect` code
+        // If you want to experiment with relaying using your own relay server,
+        // you must pass in the same custom relay url to both the `listen` code AND the `connect` code
         .relay_mode(RelayMode::Default)
-        // you can choose a port to bind to, but passing in `0` will bind the socket to a random available port
+        // pass in `0` to bind the socket to a random available port
         .bind(0)
         .await?;
 
@@ -345,7 +360,7 @@ fn get_service_manager() -> Result<Box<dyn ServiceManager>, io::Error> {
     Ok(manager)
 }
 
-// This will install the binary as a user level service and then start it
+/// Install the binary as a user level service and then start it
 fn install_service(service_name: &ServiceLabel, exec_path: PathBuf) -> Result<(), io::Error> {
     let manager = get_service_manager()?;
     // Run from dir where exec is for now, so it should find the config file in ancestors path
@@ -383,7 +398,7 @@ fn install_service(service_name: &ServiceLabel, exec_path: PathBuf) -> Result<()
     Ok(())
 }
 
-// this will stop any running instance of the service, then uninstall it
+/// Stop any running instance of the service, then uninstall it
 fn uninstall_service(service_name: &ServiceLabel) -> Result<(), io::Error> {
     let manager = get_service_manager()?;
 

--- a/src/piglet.rs
+++ b/src/piglet.rs
@@ -104,7 +104,7 @@ fn check_unique(exec_path: &Path) -> anyhow::Result<PathBuf> {
     let sys = System::new_all();
     let instances: Vec<&Process> = sys
         .processes_by_exact_name(exec_name)
-        .filter(|p| p.pid().as_u32() != my_pid)
+        .filter(|p| p.thread_kind().is_none() && p.pid().as_u32() != my_pid)
         .collect();
     if let Some(process) = instances.first() {
         println!(

--- a/src/piglet.rs
+++ b/src/piglet.rs
@@ -8,7 +8,8 @@ use futures_lite::StreamExt;
 use hw::config::HardwareConfig;
 use hw::Hardware;
 use iroh_net::endpoint::Connection;
-use iroh_net::{key::SecretKey, relay::RelayMode, Endpoint};
+use iroh_net::relay::RelayUrl;
+use iroh_net::{key::SecretKey, relay::RelayMode, Endpoint, NodeId};
 use log::error;
 use log::{info, trace};
 use service_manager::{
@@ -16,11 +17,13 @@ use service_manager::{
     ServiceUninstallCtx,
 };
 use std::env::current_exe;
-use std::path::PathBuf;
+use std::fs::File;
+use std::io::Write;
+use std::path::{Path, PathBuf};
 use std::process::exit;
 use std::str::FromStr;
 use std::time::Duration;
-use std::{env, io, process};
+use std::{env, fs, io, process};
 use sysinfo::{Process, System};
 use tracing::Level;
 use tracing_subscriber::filter::{Directive, LevelFilter};
@@ -35,30 +38,23 @@ const SERVICE_NAME: &str = "net.mackenzie-serres.pigg.piglet";
 #[tokio::main]
 async fn main() -> anyhow::Result<()> {
     let matches = get_matches();
+    let exec_path = current_exe()?;
 
-    manage_service(&matches)?;
+    manage_service(&exec_path, &matches)?;
 
-    run_service(&matches).await
+    let info_path = check_unique(&exec_path)?;
+
+    run_service(&info_path, &matches).await
 }
 
 /// Handle any service installation or uninstallation tasks
-fn manage_service(matches: &ArgMatches) -> anyhow::Result<()> {
+fn manage_service(exec_path: &Path, matches: &ArgMatches) -> anyhow::Result<()> {
     let service_name: ServiceLabel = SERVICE_NAME.parse().unwrap();
 
     if matches.get_flag("uninstall") {
         uninstall_service(&service_name)?;
         exit(0);
     }
-
-    let exec_path = current_exe()?;
-
-    check_unique(
-        exec_path
-            .file_name()
-            .context("Could not get exec file name")?
-            .to_str()
-            .context("Could not get exec file name")?,
-    )?;
 
     if matches.get_flag("install") {
         install_service(&service_name, exec_path)?;
@@ -70,7 +66,7 @@ fn manage_service(matches: &ArgMatches) -> anyhow::Result<()> {
 
 /// Run piglet as a service - this could be interactively by a user in foreground or started
 /// by the system as a user service, in background - use logging for output from here on
-async fn run_service(matches: &ArgMatches) -> anyhow::Result<()> {
+async fn run_service(info_path: &Path, matches: &ArgMatches) -> anyhow::Result<()> {
     setup_logging(matches);
 
     let mut hw = hw::get();
@@ -88,7 +84,7 @@ async fn run_service(matches: &ArgMatches) -> anyhow::Result<()> {
     };
 
     // Then listen for remote connections and "serve" them
-    listen(hw).await
+    listen(info_path, hw).await
 }
 
 /// CHeck that this is the only instance of piglet running, both user process or system process
@@ -96,7 +92,13 @@ async fn run_service(matches: &ArgMatches) -> anyhow::Result<()> {
 /// - print out that fact, with the process ID
 /// - print out the nodeid of the instance that is running
 /// - exit
-fn check_unique(exec_name: &str) -> anyhow::Result<()> {
+fn check_unique(exec_path: &Path) -> anyhow::Result<PathBuf> {
+    let exec_name = exec_path
+        .file_name()
+        .context("Could not get exec file name")?
+        .to_str()
+        .context("Could not get exec file name")?;
+
     let my_pid = process::id();
     let sys = System::new_all();
     let instances: Vec<&Process> = sys
@@ -114,7 +116,11 @@ fn check_unique(exec_name: &str) -> anyhow::Result<()> {
         exit(1);
     }
 
-    Ok(())
+    let info_path = exec_path.with_file_name("piglet.info");
+    // remove any leftover file from a previous execution - ignore any failure
+    let _ = fs::remove_file(&info_path);
+
+    Ok(info_path)
 }
 
 /// Setup logging with the requested verbosity level - or default if none was specified
@@ -180,7 +186,7 @@ fn get_matches() -> ArgMatches {
 /// Listen for an incoming iroh-net connection and apply any config changes received, and
 /// send to GUI over the connection any input level changes.
 /// This is adapted from the iroh-net example with help from the iroh community
-async fn listen(mut hardware: impl Hardware) -> anyhow::Result<()> {
+async fn listen(info_path: &Path, mut hardware: impl Hardware) -> anyhow::Result<()> {
     let secret_key = SecretKey::generate();
 
     // Build a `Endpoint`, which uses PublicKeys as node identifiers, uses QUIC for directly
@@ -204,8 +210,8 @@ async fn listen(mut hardware: impl Hardware) -> anyhow::Result<()> {
         .bind(0)
         .await?;
 
-    let me = endpoint.node_id();
-    info!("node id: {me}");
+    let nodeid = endpoint.node_id();
+    info!("node id: {nodeid}");
 
     let local_addrs = endpoint
         .direct_addresses()
@@ -222,6 +228,9 @@ async fn listen(mut hardware: impl Hardware) -> anyhow::Result<()> {
         .home_relay()
         .expect("should be connected to a relay server, try calling `endpoint.local_endpoints()` or `endpoint.connect()` first, to ensure the endpoint has actually attempted a connection before checking for the connected relay server");
     info!("node relay server url: {relay_url}");
+
+    // write the info about the node to the info_path file for use in piggui
+    write_info_file(info_path, &nodeid, &local_addrs, &relay_url)?;
 
     // accept incoming connections, returns a normal QUIC connection
     if let Some(connecting) = endpoint.accept().await {
@@ -255,6 +264,21 @@ async fn listen(mut hardware: impl Hardware) -> anyhow::Result<()> {
         }
     }
 
+    Ok(())
+}
+
+/// Write info about the running piglet to the info file
+fn write_info_file(
+    info_path: &Path,
+    nodeid: &NodeId,
+    local_addrs: &str,
+    relay_url: &RelayUrl,
+) -> anyhow::Result<()> {
+    let mut output = File::create(info_path)?;
+    writeln!(output, "nodeid:{nodeid}")?;
+    writeln!(output, "local addresses:{local_addrs}")?;
+    writeln!(output, "relay_url:{relay_url}")?;
+    info!("Info file written at: {info_path:?}");
     Ok(())
 }
 
@@ -361,7 +385,7 @@ fn get_service_manager() -> Result<Box<dyn ServiceManager>, io::Error> {
 }
 
 /// Install the binary as a user level service and then start it
-fn install_service(service_name: &ServiceLabel, exec_path: PathBuf) -> Result<(), io::Error> {
+fn install_service(service_name: &ServiceLabel, exec_path: &Path) -> Result<(), io::Error> {
     let manager = get_service_manager()?;
     // Run from dir where exec is for now, so it should find the config file in ancestors path
     let exec_dir = exec_path
@@ -375,7 +399,7 @@ fn install_service(service_name: &ServiceLabel, exec_path: PathBuf) -> Result<()
     // Install our service using the underlying service management platform
     manager.install(ServiceInstallCtx {
         label: service_name.clone(),
-        program: exec_path.clone(),
+        program: exec_path.to_path_buf(),
         args: vec![],
         contents: None, // Optional String for system-specific service content.
         username: None, // Optional String for alternative user to run service.

--- a/src/piglet.rs
+++ b/src/piglet.rs
@@ -11,13 +11,23 @@ use iroh_net::endpoint::Connection;
 use iroh_net::{key::SecretKey, relay::RelayMode, Endpoint};
 use log::error;
 use log::{info, trace};
+use service_manager::{
+    ServiceInstallCtx, ServiceLabel, ServiceManager, ServiceStartCtx, ServiceStopCtx,
+    ServiceUninstallCtx,
+};
+use std::env::current_exe;
+use std::path::PathBuf;
+use std::process::exit;
 use std::str::FromStr;
-use std::{env, io};
+use std::time::Duration;
+use std::{env, io, process};
+use sysinfo::{Process, System};
 use tracing::Level;
 use tracing_subscriber::filter::{Directive, LevelFilter};
 use tracing_subscriber::EnvFilter;
 
 mod hw;
+const SERVICE_NAME: &str = "net.mackenzie-serres.pigg.piglet";
 
 /// Piglet will expose the same functionality from the GPIO Hardware Backend used by the GUI
 /// in Piggy, but without any GUI or related dependencies, loading a config from file and
@@ -30,8 +40,30 @@ async fn main() -> anyhow::Result<()> {
 
 /// Do initialization of logging, get access to local hardware and apply any config from a local
 /// file that maybe specified on the command line
-fn init() -> io::Result<impl Hardware> {
+fn init() -> anyhow::Result<impl Hardware> {
     let matches = get_matches();
+
+    let service_name: ServiceLabel = SERVICE_NAME.parse().unwrap();
+
+    if matches.get_flag("uninstall") {
+        uninstall_service(&service_name)?;
+        exit(0);
+    }
+
+    let exec_path = current_exe()?;
+
+    check_unique(
+        exec_path
+            .file_name()
+            .context("Could not get exec file name")?
+            .to_str()
+            .context("Could not get exec file name")?,
+    )?;
+
+    if matches.get_flag("install") {
+        install_service(&service_name, exec_path)?;
+        exit(0);
+    };
 
     setup_logging(&matches);
 
@@ -50,6 +82,32 @@ fn init() -> io::Result<impl Hardware> {
     };
 
     Ok(hw)
+}
+
+/// CHeck that this is the only instance of piglet running, both user process or system process
+/// If another version is detected:
+/// - print out that fact, with the process ID
+/// - print out the nodeid of the instance that is running
+/// - exit
+fn check_unique(exec_name: &str) -> anyhow::Result<()> {
+    let my_pid = process::id();
+    let sys = System::new_all();
+    let instances: Vec<&Process> = sys
+        .processes_by_exact_name(exec_name)
+        .filter(|p| p.pid().as_u32() != my_pid)
+        .collect();
+    if let Some(process) = instances.first() {
+        println!(
+            "An instance of {exec_name} is already running with PID='{}' with Path='{}', started by user with {:?}",
+            process.pid(),
+            process.exe().context("Could not get path to the running process instance")?.display(),
+            process.user_id().context("Could not get the User ID of user who started the running process instance")?,
+        );
+
+        exit(1);
+    }
+
+    Ok(())
 }
 
 /// Setup logging with the requested verbosity level - or default if none was specified
@@ -71,6 +129,24 @@ fn get_matches() -> ArgMatches {
 
     let app = app.about(
         "'piglet' - for making Raspberry Pi GPIO hardware accessible remotely using 'piggui'",
+    );
+
+    let app = app.arg(
+        Arg::new("install")
+            .short('i')
+            .long("install")
+            .action(clap::ArgAction::SetTrue)
+            .help("Install piglet as a System Service that restarts on reboot")
+            .conflicts_with("uninstall"),
+    );
+
+    let app = app.arg(
+        Arg::new("uninstall")
+            .short('u')
+            .long("uninstall")
+            .action(clap::ArgAction::SetTrue)
+            .help("Uninstall any piglet System Service")
+            .conflicts_with("install"),
     );
 
     let app = app.arg(
@@ -258,5 +334,76 @@ async fn send(connection: Connection, message: String) -> anyhow::Result<()> {
     let mut gui_sender = connection.open_uni().await?;
     gui_sender.write_all(message.as_bytes()).await?;
     gui_sender.finish().await?;
+    Ok(())
+}
+
+fn get_service_manager() -> Result<Box<dyn ServiceManager>, io::Error> {
+    // Get generic service by detecting what is available on the platform
+    let manager = <dyn ServiceManager>::native()
+        .map_err(|_| io::Error::new(io::ErrorKind::NotFound, "Could not create ServiceManager"))?;
+
+    Ok(manager)
+}
+
+// This will install the binary as a user level service and then start it
+fn install_service(service_name: &ServiceLabel, exec_path: PathBuf) -> Result<(), io::Error> {
+    let manager = get_service_manager()?;
+    // Run from dir where exec is for now, so it should find the config file in ancestors path
+    let exec_dir = exec_path
+        .parent()
+        .ok_or(io::Error::new(
+            io::ErrorKind::NotFound,
+            "Could not get exec dir",
+        ))?
+        .to_path_buf();
+
+    // Install our service using the underlying service management platform
+    manager.install(ServiceInstallCtx {
+        label: service_name.clone(),
+        program: exec_path.clone(),
+        args: vec![],
+        contents: None, // Optional String for system-specific service content.
+        username: None, // Optional String for alternative user to run service.
+        working_directory: Some(exec_dir),
+        environment: None, // Optional list of environment variables to supply the service process.
+        autostart: true,
+    })?;
+
+    // Start our service using the underlying service management platform
+    manager.start(ServiceStartCtx {
+        label: service_name.clone(),
+    })?;
+
+    println!(
+        "'service '{}' ('{}') installed and started",
+        service_name,
+        exec_path.display()
+    );
+
+    Ok(())
+}
+
+// this will stop any running instance of the service, then uninstall it
+fn uninstall_service(service_name: &ServiceLabel) -> Result<(), io::Error> {
+    let manager = get_service_manager()?;
+
+    // Stop our service using the underlying service management platform
+    manager.stop(ServiceStopCtx {
+        label: service_name.clone(),
+    })?;
+
+    println!(
+        "service '{}' stopped. Waiting for 10s before uninstalling",
+        service_name
+    );
+    std::thread::sleep(Duration::from_secs(10));
+
+    // Uninstall our service using the underlying service management platform
+    manager.uninstall(ServiceUninstallCtx {
+        label: service_name.clone(),
+    })?;
+
+    println!("service '{}' uninstalled", service_name);
+
     Ok(())
 }

--- a/src/piglet.rs
+++ b/src/piglet.rs
@@ -236,39 +236,47 @@ async fn listen(info_path: &Path, mut hardware: impl Hardware) -> anyhow::Result
     // write the info about the node to the info_path file for use in piggui
     write_info_file(info_path, &nodeid, &local_addrs, &relay_url)?;
 
-    // accept incoming connections, returns a normal QUIC connection
-    if let Some(connecting) = endpoint.accept().await {
-        let connection = connecting.await?;
-        let node_id = iroh_net::endpoint::get_remote_node_id(&connection)?;
-        info!("New connection from nodeid: '{node_id}'",);
+    loop {
+        // accept incoming connections, returns a normal QUIC connection
+        if let Some(connecting) = endpoint.accept().await {
+            let connection = connecting.await?;
+            let node_id = iroh_net::endpoint::get_remote_node_id(&connection)?;
+            info!("New connection from nodeid: '{node_id}'",);
 
-        let mut gui_sender = connection.open_uni().await?;
+            let mut gui_sender = connection.open_uni().await?;
 
-        trace!("Sending hardware description");
-        let desc = hardware.description()?;
-        let message = serde_json::to_string(&desc)?;
-        gui_sender.write_all(message.as_bytes()).await?;
-        gui_sender.finish().await?;
+            trace!("Sending hardware description");
+            let desc = hardware.description()?;
+            let message = serde_json::to_string(&desc)?;
+            gui_sender.write_all(message.as_bytes()).await?;
+            gui_sender.finish().await?;
 
-        loop {
-            trace!("waiting for connection");
-            let mut config_receiver = connection.accept_uni().await?;
-            let connection_clone = connection.clone();
-            trace!("Connected, waiting for message");
-            let payload = config_receiver.read_to_end(4096).await?;
+            loop {
+                trace!("waiting for connection");
+                match connection.accept_uni().await {
+                    Ok(mut config_receiver) => {
+                        let connection_clone = connection.clone();
+                        trace!("Connected, waiting for message");
+                        let payload = config_receiver.read_to_end(4096).await?;
 
-            if !payload.is_empty() {
-                let content = String::from_utf8_lossy(&payload);
-                if let Ok(config_message) = serde_json::from_str(&content) {
-                    apply_config_change(&mut hardware, config_message, connection_clone).await
-                } else {
-                    error!("Unknown message: {content}");
-                };
+                        if !payload.is_empty() {
+                            let content = String::from_utf8_lossy(&payload);
+                            if let Ok(config_message) = serde_json::from_str(&content) {
+                                apply_config_change(&mut hardware, config_message, connection_clone)
+                                    .await
+                            } else {
+                                error!("Unknown message: {content}");
+                            };
+                        }
+                    }
+                    _ => {
+                        info!("Connection lost");
+                        break;
+                    }
+                }
             }
         }
     }
-
-    Ok(())
 }
 
 /// Write info about the running piglet to the info file
@@ -279,9 +287,9 @@ fn write_info_file(
     relay_url: &RelayUrl,
 ) -> anyhow::Result<()> {
     let mut output = File::create(info_path)?;
-    writeln!(output, "nodeid:{nodeid}")?;
-    writeln!(output, "local addresses:{local_addrs}")?;
-    writeln!(output, "relay_url:{relay_url}")?;
+    writeln!(output, "nodeid : {nodeid}")?;
+    writeln!(output, "local addresses : {local_addrs}")?;
+    writeln!(output, "relay_url : {relay_url}")?;
     info!("Info file written at: {info_path:?}");
     Ok(())
 }

--- a/src/piglet.rs
+++ b/src/piglet.rs
@@ -108,17 +108,19 @@ fn check_unique(exec_path: &Path) -> anyhow::Result<PathBuf> {
         .collect();
     if let Some(process) = instances.first() {
         println!(
-            "An instance of {exec_name} is already running with PID='{}',  Path='{}'",
+            "An instance of {exec_name} is already running with PID='{}'",
             process.pid(),
-            process
-                .exe()
-                .context("Could not get path to the running process instance")?
-                .display(),
         );
 
-        println!("You can use the following info to connect to this running instance:");
-        let piglet_info = fs::read_to_string(info_path)?;
-        println!("{}", piglet_info);
+        // If we can find the path to the executable - look for the info file
+        if let Some(path) = process.exe() {
+            let info_path = path.with_file_name("piglet.info");
+            if info_path.exists() {
+                println!("You can use the following info to connect to this running instance:");
+                let piglet_info = fs::read_to_string(info_path)?;
+                println!("{}", piglet_info);
+            }
+        }
 
         exit(1);
     }

--- a/src/piglet.rs
+++ b/src/piglet.rs
@@ -108,10 +108,12 @@ fn check_unique(exec_path: &Path) -> anyhow::Result<PathBuf> {
         .collect();
     if let Some(process) = instances.first() {
         println!(
-            "An instance of {exec_name} is already running with PID='{}',  Path='{}', started by user with {:?}",
+            "An instance of {exec_name} is already running with PID='{}',  Path='{}'",
             process.pid(),
-            process.exe().context("Could not get path to the running process instance")?.display(),
-            process.user_id().context("Could not get the User ID of user who started the running process instance")?,
+            process
+                .exe()
+                .context("Could not get path to the running process instance")?
+                .display(),
         );
 
         println!("You can use the following info to connect to this running instance:");

